chap-cheri-x86-64.tex

```diff
@@ -1260,3 +1260,300 @@ \subsection{XCHG [ER]AX Opcodes}
+could disable bounds checking for this process such that all
+bounds-check operations become NOPs.
+Thus MPX was not usable for security from a determined attacker,
+but only to accelerate
+identification of memory safety vulnerabilities in a friendly
+environment.
+
+Secondly, MPX performance overhead was high;
+over 50\% on SPEC using the Intel compiler.
+Code bloat was considerable, with new instructions
+for loading and storing bounds alongside pointers,
+and separate instructions for checking the lower
+and upper bounds.
+The table access mode could quintuple the memory footprint
+of a program, though SPEC averaged 100\% memory overhead
+due to the sparse table structure.
+Furthermore, limitations on execution units that could
+execute bounds checks meant that instructions-per-cycle
+was usually lower for a MPX program than for a standard
+program despite the instruction bloat.
+
+With no hard security guarantees and performance competitive
+With similar software schemes, MPX adoption
+was abysmal and was deprecated after one generation of
+hardware.
+
+\subsubsection{MPX Lessons for CHERI-x86-64}
+
+CHERI sacrifices ABI compatibility for firm security guarantees
+and performance.
+While the x86 community has placed great emphasis on backward
+compatibility, CHERI-x86-64 must resist the pressure
+to cleverly enable compatibility in any way that compromises
+security.
+
+In addition, CHERI-x86-64 implementors should resist the pressure
+to leave CHERI de-optimised under the assumption that security
+is a niche application.
+CHERI is designed for overheads within 5\% of unprotected
+performance in most cases, but a few corner cases in forwarding
+or branch prediction can easily push overheads over 20\% and
+greatly restrict adoption.
+
+\subsection{Intel SGX (deprecated)}
+
+Intel SGX is a very complex extension to enable a user-space
+application to trust the processor to provide enclave memory that is
+not readable or modifiable from the operating system or even
+outside the CPU package.
+Some instructions required for this extension are in the style
+of function calls.
+The ECREATE instruction takes a pointer to a struct as an argument,
+which requests a contiguous range of virtual address space.
+This causes the hardware to allocate a number of structures in a
+reserved region of DRAM to track memory for the enclave before
+executing the next instruction.
+The EADD instruction copies a page of common memory into a
+newly-allocated enclave page, keeping a running hash of its contents.
+
+CHERI relies on simple invariants that can be enforced within
+single-cycle instructions.
+Intel SGX instead trusts the correctness and atomicity of
+hyper-visor-like subroutine calls due to their being
+provided by processor firmware.
+
+On one hand, the SGX precedent suggests that trusted hyper-visor-like
+subroutines can be implemented in the CPU.
+On the other hand, Intel SGX was only implemented in a single Intel
```

Comment:
I'm not sure these next two sentences are accurate.  Intel's current top-of-line Xeon processors ship SGX: https://www.intel.com/content/www/us/en/products/details/processors/xeon/scalable/platinum.html